---
title: "Introduction to TreeSE: interactive visualization for Single Cell RNA seq data"
author: "HÃ©ctor Corrada Bravo, Jayaram Kancherla, Kazi Tasnim Zinat"
date: "`r Sys.Date()`"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Introduction to TreeSE}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

[`TreeSE`](https://github.com/HCBravoLab/TreeSE) is a tool for interactive visualization and exploration of Single Cell RNA sequencing data. TreeSE provides a way for exploring clustered feature data hierarchically, 
while supporting other useful data visualization charts like heatmaps.

# Seurat example

To demonstrate the functionality we will work through a simple
example using the `pbmc` dataset.

```{r, eval=TRUE, echo=TRUE, results='hide', warning=FALSE, error=FALSE, message=FALSE}
library(scrnaviz)
library(dplyr)
library(Seurat)
library(patchwork)
```

```{r}
#Create PBMC and clustree
runSeurat<- function(datadir){
  pbmc.data <- Read10X(data.dir = datadir)
  pbmc <- CreateSeuratObject(counts = pbmc.data, min.cells = 3, min.features = 200,
                             project = "pbmc3k")
  pbmc
  pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")
  pbmc <- NormalizeData(pbmc)
  all.genes <- rownames(pbmc)
  pbmc <- ScaleData(pbmc, vars.to.regress = "percent.mt")
  pbmc <- FindVariableFeatures(object = pbmc)
  pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))
  pbmc <- FindNeighbors(pbmc, dims = 1:10)
  pbmc <- FindClusters(pbmc, resolution = c(0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0), print.output = 0, save.SNN = TRUE)
  pbmc
}
```

```{r, eval=TRUE, echo=TRUE, results='hide', warning=FALSE, error=FALSE, message=FALSE}
#Run SEURAT
pbmc<- runSeurat("../data/filtered_gene_bc_matrices/hg19/")
#filter cluster data
clusterdata <- pbmc@meta.data
clusterdata <- clusterdata %>%
  select(starts_with("RNA_snn"))
```
```{r, eval=TRUE, echo=TRUE, results='hide', warning=FALSE, error=FALSE, message=FALSE}
pbmc_TreeSE <-
  reassign_and_collapse(clusterdata, GetAssayData(pbmc))
#app <- startMetaviz()
```

