---
title: "Introduction to TreeSE: interactive visualization for Single Cell RNA seq data"
author: "HÃ©ctor Corrada Bravo, Jayaram Kancherla, Kazi Tasnim Zinat"
date: "`r Sys.Date()`"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Introduction to TreeSE}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction

[`TreeSE`](https://github.com/HCBravoLab/TreeSE) is a tool for interactive visualization and exploration of Single Cell RNA sequencing data. TreeSE provides a way for exploring clustered feature data hierarchically, 
while supporting other useful data visualization charts like heatmaps.

# Simple Example

Given a `DataFrame` with cluster information, and a count matrix, a `TreeSE` object is generated the following way
```{r, eval=TRUE, echo=TRUE, results='hide', warning=FALSE, error=FALSE, message=FALSE}
n=64
df<- data.frame(cluster0=rep(1,n))
for(i in seq(1,5)){
  df[[paste0("cluster",i)]]<- rep(seq(1:(2**i)),each=ceiling(n/(2**i)),len=n)
}
counts <- matrix(rpois(6400, lambda = 10), ncol=n, nrow=100)
TreeSE<-simplified_treese(df,counts)
```

# Preprocessing

Existing popular Single Cell packages can provide the necessary clustering information required to visualize the object. Here, we show the pipeline for using `TreeSE` with two widely used Single Cell RNA-seq packages, `Seurat` and `SingleCellExperiment`.

## `Seurat` example

To demonstrate the functionality we will work through a simple
example using the `pbmc` dataset. 

### Pre-processing raw data and Finding CLusters

First, we create a Seurat object, and run clustering at various resolutions (0-1.0)

```{r, eval=TRUE, echo=TRUE, results='hide', warning=FALSE, error=FALSE, message=FALSE}
library(scrnaviz)
library(dplyr)
library(Seurat)
```

The following function converts the data to Seurat object and finds clusters at distinct resolutions.
```{r}
runSeurat<- function(){
  pbmc <- pbmc_small
  pbmc
  pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")
  pbmc <- NormalizeData(pbmc)
  all.genes <- rownames(pbmc)
  pbmc <- ScaleData(pbmc, vars.to.regress = "percent.mt")
  pbmc <- FindVariableFeatures(object = pbmc)
  pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))
  pbmc <- FindNeighbors(pbmc, dims = 1:10)
  pbmc <- FindClusters(pbmc, resolution = c(0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0), print.output = 0, save.SNN = TRUE)
  pbmc
}
```


### Filtering out the Cluster Information

Once we have the clusters, we get the clusterdata from the Seurat object metadata and select only the columns that provide us clustering information. This step gives us the dataframe with relevant cluster information.
```{r, eval=TRUE, echo=TRUE, results='hide', warning=FALSE, error=FALSE, message=FALSE}
#Run SEURAT
pbmc<- runSeurat()
#filter cluster data
clusterdata <- pbmc@meta.data
clusterdata <- clusterdata %>%
  select(starts_with("RNA_snn"))
```

### Get the Tree Summarized Experiment bject to be visualized

Now, we provide `TreeSE` class with the `Seurat` object. In return we get the `TreeSE` result ready for visualization.
```{r, eval=TRUE, echo=TRUE,  warning=FALSE, error=FALSE, message=FALSE}
#Run Seurat
TreeSE<-visualizeSeurat(pbmc)
head(TreeSE)
```


### Getting the most variable genes

For HeatMap visualization, we first extract the top 100 most variable genes from the samples.
```{r, eval=TRUE, echo=TRUE}
TreeSE<-find_top_variable_genes(TreeSE,100)
```

## `SingleCellExperiment` Example

To demonstrate the functionality we will work through the `Mouse mebryonic stem cells`  dataset available via `scRNAseq` package, which provides access to a collection of public scRNA-seq dasets in the form of `SingleCellExperiment` object. 


```{r, eval=TRUE, echo=TRUE, results='hide', warning=FALSE, error=FALSE, message=FALSE}
library(scRNAseq)
library(SC3)
library(scran)
library(scater)
```

The following code generates the `SingleCellExperiment` object.

```{r, eval=TRUE, echo=TRUE, results='hide', warning=FALSE, error=FALSE, message=FALSE}

sce<- mockSCE()

# Normalization.
sce <- logNormCounts(sce)


# define feature names in feature_symbol column
rowData(sce)$feature_symbol <- rownames(sce)
# remove features with duplicated names
sce <- sce[!duplicated(rowData(sce)$feature_symbol), ]

# Dimensionality reduction.

sce <- runPCA(sce)

```

### Creating clusters

We then create clusters at different resolutions using the `SC3` package.

```{r, eval=FALSE, echo=TRUE, results='hide', warning=FALSE, error=FALSE, message=FALSE}
counts(sce) <- as.matrix(counts(sce))
#normcounts(sce) <- as.matrix(normcounts(sce))
logcounts(sce) <- as.matrix(logcounts(sce))
#sce <- sc3_kmeans(sce, ks = 1:10)
sce <- sc3(sce, ks = 1:10)
```

```{r, eval=TRUE, echo=TRUE, results='hide', warning=FALSE, error=FALSE, message=FALSE}
for(i in  seq(10)){
  clust.kmeans <- kmeans(reducedDim(sce, "PCA"), centers=i)
  sce[[paste0("cluster",i)]] <- factor(clust.kmeans$cluster)
  
}
```


### Filtering out the Cluster Information

Once we have the clusters, we get the clusterdata from the Seurat object metadata and select only the columns that provide us clustering information. This step gives us the dataframe with relevant cluster information.

```{r, eval=FALSE, echo=TRUE, results='hide', warning=FALSE, error=FALSE, message=FALSE}

#as.vector(counts(sce))
col_data <- colData(sce)
col_data <- col_data[ , grep("sc3_", colnames(col_data))]
```

## Get the Tree Summarized Experiment object to be visualized

Now, we provide `TreeSE` class with the cluster dataframe and associated count matrix. In return we get the `TreeSE` result ready for visualization.
```{r, eval=TRUE, echo=TRUE,  warning=FALSE, error=FALSE, message=FALSE}
TreeSE<-visualizeSingleCellExperiment(sce)
head(TreeSE)
```

### Getting the most variable genes

For HeatMap visualization, we first extract the top 100 most variable genes from the   samples.
```{r, eval=TRUE, echo=TRUE,  warning=FALSE, error=FALSE, message=FALSE}
TreeSE<-find_top_variable_genes(TreeSE,100)
```

# Interactive visualization

## The metavizr session manager

In the next step, we visualize the interactive tree via 'Metaviz' app, by calling `startMetaviz` function.
```{r, eval=FALSE, echo=TRUE}
app <- startMetaviz()
```

This opens a websocket connection between the interactive `R` session and the browser client. This will allow us to visualize data stored in the `Metaviz` server along with data in the interactive `R` session.

## Plotting the data

By calling the following plot function we plot the 'TreeSE' object via Metaviz
```{r, eval=FALSE, echo=TRUE}
icicle_plot <-
  app$plot(TreeSE, datasource_name = "SCRNA", tree = "col")
```
## Visualizing the heatMap


We then get the relevant object list from icicle plot and based on the count matrix from plot, subset the 100 most variable genes
```{r, eval=FALSE, echo=TRUE}
facetZoom <- app$get_ms_object(chart_id_or_object = icicle_plot)
# Get Measurements from the plot
ms_list <- facetZoom$get_measurements()
subset_ms_list <- Filter(function(ms)
  ms@id %in% metadata(TreeSE)$top_variable, ms_list)
```
### Visualization via Metaviz

The following command adds a HeatMap plot in the browser.
```{r, eval=FALSE, echo=TRUE}
app$chart_mgr$visualize(chart_type = "HeatmapPlot", measurements = subset_ms_list)
```
